servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/DefaultHost.java

```diff
@@ -624,20 +531,95 @@ public String toString() {
         }
     }
 
-    private static final class ConnState {
-        final Object[] connections;
-        final Object state;
-
-        ConnState(final Object[] connections, final Object state) {
+    private final class ConnState {
+
+        final List<C> connections;
+        final State state;
+        final int failedConnections;
+        @Nullable
+        HealthCheck healthCheck;
+
+        private ConnState(final List<C> connections, State state, int failedConnections, HealthCheck healthCheck) {
+            // These asserts codify the invariants of the state.
+            // if the state is unhealthy there must be a healthcheck
+            assert state != State.UNHEALTHY || healthCheck != null;
+            // if the state is not unhealthy then there must not be a healthcheck.
+            assert state == State.UNHEALTHY || healthCheck == null;
+            // Only unhealthy and active states should be counting failed connections.
+            assert (state == State.UNHEALTHY || state == State.ACTIVE) || failedConnections == 0;
             this.connections = connections;
             this.state = state;
+            this.failedConnections = failedConnections;
+            this.healthCheck = healthCheck;
+        }
+
+        ConnState toNextFailedConnection(Throwable cause) {
```

Comment:
nit: final (other places too)

Reply:
The class is final so adding final to the methods triggers checkstyle redundancy errors.